python/pyspark/sql/connect/plan.py

```diff
@@ -721,8 +721,19 @@ def plan(self, session: "SparkConnectClient") -> proto.Relation:
                     rel.join.using_columns.append(self.on)
                 else:
                     rel.join.join_condition.CopyFrom(self.to_attr_or_expression(self.on, session))
-            else:
-                rel.join.using_columns.extend(self.on)
+            elif len(self.on) > 0:
+                if isinstance(self.on[0], str):
+                    rel.join.using_columns.extend(self.on)  # ignore: arg-type
+                else:
+                    merge_column = None
+                    for c in self.on:
+                        if merge_column is not None:
+                            merge_column = merge_column & c
+                        else:
+                            merge_column = c
+                    rel.join.join_condition.CopyFrom(
+                        merge_column.to_plan(session)  # ignore: union-attr
```

Comment:
Same for this.

Reply:
```
python/pyspark/sql/connect/plan.py:735: error: Item "str" of "Union[str, Column, None]" has no attribute "to_plan"  [union-attr]
python/pyspark/sql/connect/plan.py:735: error: Item "None" of "Union[str, Column, None]" has no attribute "to_plan"  [union-attr]
```

Reply:
So here the challenge is: if self.on is empty, merge_column is None (Optional). On itself is defined as either str, Column or None, so when you asign it might be any of those.

My suggestion would be to add proper isinstance checks or do a manual conversion to the expected type.

Reply:
With your analysis I see the error better now. 

I chose to cast merged_column to `Column` which fix the error because I did `len(self.on) > 0` first then also rule out the str case by `isinstance(self.on[0], str)`.